oeps/oep-0050-hooks-extension-framework.rst

```diff
@@ -0,0 +1,235 @@
+themes. Given that Open edX already has a variety of theming capabilities and
+options we have opted here not to include any hooks that are meant to be used
+for theming.
+
+This proposal does not copy the model verbatim, since the open edX platform uses
+a different base technology which already allows for many extension points.
+This proposal deals primarily with the specific spots in the code execution that
+for many initiatives have been identified as the critical points that would
+allow them to adopt this technology without requiring modifications to the core.
+
+On the technical side, Django and the support for auto-installed plugins using
+stevedore already allow the addition of code in a very clean way. This proposal
+builds on top of this, only trying to standardize and grow the current
+capabilities of the platform.
+
+.. _two types of hooks: https://developer.wordpress.org/plugins/hooks/#actions-vs-filters
+
+
+Performance Considerations
+--------------------------
+
+This proposal is a double-edged sword. By allowing plugins to install code on so
+many and so varied places of the application performance could be completely
+destroyed to the point of making the platform unusable.
+
+This is a very important topic, but we believe that any responsible
+implementation of a hook must keep its own consequences and risks in check.
+In other words, it is up to the users of the hooks framework to be mindful of
+the performance of the platform by connecting to the appropriate triggers and
+optimizing their actions to the appropriate scale their Open edX instance runs on.
+
+Using asynchronous tasks to fulfill the actions and moreover connecting actions
+to the triggers in a way that is directly asynchronous is strongly recommended.
+The initial implementation must include sufficient tooling to make this possible
+and encourage it.
+
+Alternatively, in the core platform code, the actual exposition of each
+particular trigger could be configured (via flags or settings) so that even for
+scales at which the simple lookup for the actions that might be connected to a
+particular trigger could have performance implications is accounted for.
+
+
+Backward Compatibility
+======================
+
+The only extension point that will be affected by this OEP is the
+REGISTRATION_EXTENSION_FORM. This extension point should still be supported
+until a clear migration path to the corresponding hook is well defined and
+documented.
+
+
+Reference Implementation
+========================
+
+The reference implementation must be completed before this OEP is given "Final"
+status. This document will be updated as the implementation progresses.
+
+
+Rejected Alternatives
+=====================
+
+* Django Signals was initially considered as the primary mechanism to connect
+triggers and actions, but was ultimately discarded due to the lack of control
+of the execution order when more than one action is listening for a trigger.
```

Comment:
[seeking more information] Django Signals are the standard Django web framework capability for this type of extension. My initial inclination would be to leverage this capability that already exists in our framework, rather than building and maintaining our own.

What use cases do we have that require a pre-defined execution order?
Also, see my question above since it's not clear how this proposal will satisfy this requirement?

Reply:
Now that django signals are back as the mechanism for supporting event hooks, the strict execution order is only there because filters need a way to 'add' or 'reduce' the result of all the functions that were registered.